src/mpid/ch4/netmod/ofi/ofi_coll_select_container_parsers.c

```diff
@@ -0,0 +1,114 @@
+
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2016 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+
+#include <mpidimpl.h>
+#include "ofi_noinline.h"
+
+const char *MPIDU_SELECTION_Bcast_ofi[] = {
+    "binomial",
+    "scatter_recursive_doubling_allgather",
+    "scatter_ring_allgatherv",
+};
+
+int MPIDU_SELECTION_MAX_OFI_ALGORITHM_INDICES[MPIDU_SELECTION_COLLECTIVES_MAX]
+    = {
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    sizeof(MPIDU_SELECTION_Bcast_ofi) / sizeof(char *),
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+};
+
+const char **MPIDU_SELECTION_OFI_ALGORITHMS[] = {
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    MPIDU_SELECTION_Bcast_ofi,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+};
+
+void MPIDI_OFI_algorithm_parser(MPIDU_SELECTION_coll_id_t coll_id, int *cnt_num,
+                                MPIDIG_coll_algo_generic_container_t * cnt, char *value)
+{
+    int count = *cnt_num, ofi_algo_id = 0;
+    for (ofi_algo_id = 0;
+         ofi_algo_id < MPIDU_SELECTION_MAX_OFI_ALGORITHM_INDICES[MPIDU_SELECTION_BCAST];
+         ofi_algo_id++) {
+        /* Find if the algorithm name specified matches any of the existing ofi algorithm */
+        if (strcasecmp(MPIDU_SELECTION_OFI_ALGORITHMS[MPIDU_SELECTION_BCAST][ofi_algo_id],
+                       value) == 0) {
+            break;
+        }
+    }
+    if (ofi_algo_id < MPIDU_SELECTION_MAX_OFI_ALGORITHM_INDICES[MPIDU_SELECTION_BCAST]) {
+        /* Assign the algo_id assosicated with the algorithm name (string) in the leaf node */
+        cnt[count++].id = ofi_algo_id;
+    }
+    *cnt_num = count;
+}
```

Comment:
question: is this init-time or per-call?

Reply:
init-time. We construct the leaf level containers here.